File: crack.py
import cs50
import crypt
import sys


# taking hash
def main():
    global oldHash
    if not len(sys.argv) == 2:
        print('Give me some hash')
        sys.exit()
    elif len(sys.argv) == 2:
        oldHash = sys.argv[1]

    # take salt from hash (first two characters)
    global salt
    salt = []
    for i in range(2):
        salt.append(oldHash[i])
    #print('salt is ', salt)

    # now salt is string = '50' - we need is string because cryp(string, string)
    salt = salt[0] + salt[1]

    if checkKey0() == True:
        sys.exit()
    if checkKey1() == True:
        sys.exit()
    if checkKey2() == True:
        sys.exit()
    if checkKey3() == True:
        sys.exit()


def checkNewHash(myKey):
    myKey = ''.join(myKey)
    newCheck = True
    newHash = crypt.crypt(myKey, salt)
    if oldHash == newHash:
        print(myKey)
        newCheck = True
    else:
        newCheck = False
    return newCheck

# check by 1 letter ****************************************************************


def checkKey0():
    check = False
    myKey = ['a']
    for i in range(97, 123):
        #cheking in lowercase
        myKey[0] = chr(i)
        check = checkNewHash(myKey)
        if check == True:
            break

        #cheking in uppercase
        myKey[0] = myKey[0].upper()
        check = checkNewHash(myKey)
        if check == True:
            break
    return check

# check by 2 letter ****************************************************************


def checkKey1():
    check = False
    myKey = ['a', 'b']
    for i in range(97, 123):
        for j in range(97, 123):

            # cheking both in lowercase
            myKey[0] = chr(i)
            myKey[1] = chr(j)

            check = checkNewHash(myKey)
            if check == True:
                break

            # cheking in aB uppercase
            myKey[0] = chr(i)
            myKey[1] = chr(j).upper()
            check = checkNewHash(myKey)
            if check == True:
                break

            # cheking in Ab uppercase
            myKey[0] = chr(i).upper()
            myKey[1] = chr(j)
            check = checkNewHash(myKey)
            if check == True:
                break

            # cheking in AB uppercase
            myKey[0] = chr(i).upper()
            myKey[1] = chr(j).upper()
            check = checkNewHash(myKey)
            if check == True:
                break

        if check == True:
            break
    return check

# check by 3 letter ****************************************************************


def checkKey2():
    check = False
    myKey = ['a', 'b', 'c']
    for i in range(97, 123):
        for j in range(97, 123):
            for k in range(97, 123):
                # cheking both in lowercase
                myKey[0] = chr(i)
                myKey[1] = chr(j)
                myKey[2] = chr(k)
                check = checkNewHash(myKey)
                if check == True:
                    break

                # cheking in Abc
                myKey[0] = chr(i).upper()
                myKey[1] = chr(j)
                myKey[2] = chr(k)
                check = checkNewHash(myKey)
                if check == True:
                    break

                # cheking in aBc
                myKey[0] = chr(i)
                myKey[1] = chr(j).upper()
                myKey[2] = chr(k)
                check = checkNewHash(myKey)
                if check == True:
                    break

                # cheking in abC
                myKey[0] = chr(i)
                myKey[1] = chr(j)
                myKey[2] = chr(k).upper()
                check = checkNewHash(myKey)
                if check == True:
                    break

                # cheking both in ABC
                myKey[0] = chr(i).upper()
                myKey[1] = chr(j).upper()
                myKey[2] = chr(k).upper()
                check = checkNewHash(myKey)
                if check == True:
                    break

                # cheking ABc
                myKey[0] = chr(i).upper()
                myKey[1] = chr(j).upper()
                myKey[2] = chr(k)
                check = checkNewHash(myKey)
                if check == True:
                    break

                # cheking AbC
                myKey[0] = chr(i).upper()
                myKey[1] = chr(j)
                myKey[2] = chr(k).upper()
                check = checkNewHash(myKey)
                if check == True:
                    break

                # cheking aBC
                myKey[0] = chr(i)
                myKey[1] = chr(j).upper()
                myKey[2] = chr(k).upper()
                check = checkNewHash(myKey)
                if check == True:
                    break

            if check == True:
                break
        if check == True:
            break

    return check

# check by 4 letter ****************************************************************


def checkKey3():
    check = False
    myKey = ['a', 'b', 'c', 'd']
    for i in range(97, 123):
        for j in range(97, 123):
            for k in range(97, 123):
                for t in range(97, 123):
                    # cheking abcd - all of them in lowercase
                    myKey[0] = chr(i)
                    myKey[1] = chr(j)
                    myKey[2] = chr(k)
                    myKey[3] = chr(t)
                    check = checkNewHash(myKey)
                    if check == True:
                        break

                    # cheking Abcd
                    myKey[0] = chr(i).upper()
                    myKey[1] = chr(j)
                    myKey[2] = chr(k)
                    myKey[3] = chr(t)
                    check = checkNewHash(myKey)
                    if check == True:
                        break

                    # cheking aBcd
                    myKey[0] = chr(i)
                    myKey[1] = chr(j).upper()
                    myKey[2] = chr(k)
                    myKey[3] = chr(t)
                    check = checkNewHash(myKey)
                    if check == True:
                        break

                    # cheking abCd
                    myKey[0] = chr(i)
                    myKey[1] = chr(j)
                    myKey[2] = chr(k).upper()
                    myKey[3] = chr(t)
                    check = checkNewHash(myKey)
                    if check == True:
                        break

                    # cheking abcD
                    myKey[0] = chr(i)
                    myKey[1] = chr(j)
                    myKey[2] = chr(k)
                    myKey[3] = chr(t).upper()
                    check = checkNewHash(myKey)
                    if check == True:
                        break

                    # cheking ABCD
                    myKey[0] = chr(i).upper()
                    myKey[1] = chr(j).upper()
                    myKey[2] = chr(k).upper()
                    myKey[3] = chr(t).upper()
                    check = checkNewHash(myKey)
                    if check == True:
                        break
#**********************************************************
                    # cheking ABcd
                    myKey[0] = chr(i).upper()
                    myKey[1] = chr(j).upper()
                    myKey[2] = chr(k)
                    myKey[3] = chr(t)
                    check = checkNewHash(myKey)
                    if check == True:
                        break

                    # cheking AbCd
                    myKey[0] = chr(i).upper()
                    myKey[1] = chr(j)
                    myKey[2] = chr(k).upper()
                    myKey[3] = chr(t)
                    check = checkNewHash(myKey)
                    if check == True:
                        break

                    # cheking AbcD
                    myKey[0] = chr(i).upper()
                    myKey[1] = chr(j)
                    myKey[2] = chr(k)
                    myKey[3] = chr(t).upper()
                    check = checkNewHash(myKey)
                    if check == True:
                        break

                    # cheking aBCd
                    myKey[0] = chr(i)
                    myKey[1] = chr(j).upper()
                    myKey[2] = chr(k).upper()
                    myKey[3] = chr(t)
                    check = checkNewHash(myKey)
                    if check == True:
                        break

                    # cheking aBcD
                    myKey[0] = chr(i)
                    myKey[1] = chr(j).upper()
                    myKey[2] = chr(k)
                    myKey[3] = chr(t).upper()
                    check = checkNewHash(myKey)
                    if check == True:
                        break

                    # cheking abCD
                    myKey[0] = chr(i)
                    myKey[1] = chr(j)
                    myKey[2] = chr(k).upper()
                    myKey[3] = chr(t).upper()
                    check = checkNewHash(myKey)
                    if check == True:
                        break
#************
                    # cheking ABCd
                    myKey[0] = chr(i).upper()
                    myKey[1] = chr(j).upper()
                    myKey[2] = chr(k).upper()
                    myKey[3] = chr(t)
                    check = checkNewHash(myKey)
                    if check == True:
                        break

                    # cheking aBCD
                    myKey[0] = chr(i)
                    myKey[1] = chr(j).upper()
                    myKey[2] = chr(k).upper()
                    myKey[3] = chr(t).upper()
                    check = checkNewHash(myKey)
                    if check == True:
                        break

                    # cheking AbCD
                    myKey[0] = chr(i).upper()
                    myKey[1] = chr(j)
                    myKey[2] = chr(k).upper()
                    myKey[3] = chr(t).upper()
                    check = checkNewHash(myKey)
                    if check == True:
                        break

                if check == True:
                    break
            if check == True:
                break
        if check == True:
            break

    return check


if __name__ == "__main__":
    main()

File: crack.c
#define _XOPEN_SOURCE
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <cs50.h>
#include <crypt.h>
#include <ctype.h>

char salt[3];
string oldHash;
bool checkNewHashLonely(char myKey);
bool checkNewHash(string myKey);

bool checkKey0(void);
bool checkKey1(void);
bool checkKey2(void);
bool checkKey3(void);


int main(int argc, string argv[])
{
    if (argc != 2)
    {
        printf("Just one argument after ./crack\n");
        return 1;
    }

//get a hashpass from user******************
    oldHash = argv[1];

//take salt from hash (first two characters)*********************
    for (int i = 0; i < 2; i++)
    {
        salt[i] = oldHash[i];
    }

// i dont know what is it = int l = strlen(&myKey[5]); myKey[i] != '\0'
    if (checkKey0() || checkKey1() || checkKey2() || checkKey3())
    {
        //do nothing, yes you can
    }

    return 0;
}

//*********************************************************

bool checkNewHashLonely(char myKey)
{
    bool newCheck = true;
    string newHash = "";
    newHash = crypt(&myKey, salt);
    if (strcmp(oldHash, newHash) == 0)
    {
        printf("%c\n", myKey);
        newCheck = true;
    }
    else
    {
        newCheck = false;
    }
    return newCheck;
}

bool checkNewHash(string myKey)
{
    bool newCheck = true;
    string newHash = "";
    newHash = crypt(myKey, salt);
    if (strcmp(oldHash, newHash) == 0)
    {
        printf("%s\n", myKey);
        newCheck = true;
    }
    else
    {
        newCheck = false;
    }
    return newCheck;
}



// check by 1 letter ****************************************************************
bool checkKey0(void)
{
    char myKey;
    bool check = false;

    for (int i = 97; i <= 122; i++)
    {
//cheking in lowercase
        myKey = (char) i;
        check = checkNewHashLonely(myKey);
        if (check == true)
        {
            i = 123;
        }

//cheking in uppercase
        if (check == false)
        {
            myKey = toupper(myKey);
            check = checkNewHashLonely(myKey);
            if (check == true)
            {
                i = 123;
            }
        }
    }
    return check;
}




// ************************************************ check by 2 letter ************************************************************
bool checkKey1(void)
{
    char myKey[3];
    memset(myKey, 0, sizeof(myKey));
    bool check = false;

    for (int i = 97; i <= 122; i++)
    {
        for (int j = 97; j <= 122; j++)
        {

//cheking both in lowercase
            myKey[0] = (char) i;
            myKey[1] = (char) j;
            check = checkNewHash(myKey);
            if (check == true)
            {
                i = 123;
                j = 123;
            }

//cheking in aB uppercase
            if (check == false)
            {
                myKey[0] = (char) i;
                myKey[1] = toupper((char) j);
                check = checkNewHash(myKey);
                if (check == true)
                {
                    i = 123;
                    j = 123;
                }
            }
//cheking in Ab uppercase
            if (check == false)
            {
                myKey[0] = toupper((char) i);
                myKey[1] = (char) j;
                check = checkNewHash(myKey);
                if (check == true)
                {
                    i = 123;
                    j = 123;
                }
            }
//cheking in AB uppercase
            if (check == false)
            {
                myKey[0] = toupper((char) i);
                myKey[1] = toupper((char) j);
                check = checkNewHash(myKey);
                if (check == true)
                {
                    i = 123;
                    j = 123;
                }
            }


        }
    }
    return check;
}


// ************************************************ check by 3 letter ********************************************************************
bool checkKey2(void)
{
    char myKey[4];
    memset(myKey, 0, sizeof(myKey));
    bool check = false;

    for (int i = 97; i <= 122; i++)
    {
        for (int j = 97; j <= 122; j++)
        {
            for (int k = 97; k <= 122; k++)
            {

//cheking abc - all of three in lowercase
                myKey[0] = (char) i;
                myKey[1] = (char) j;
                myKey[2] = (char) k;
                check = checkNewHash(myKey);
                if (check == true)
                {
                    i = 123;
                    j = 123;
                    k = 123;
                }

//cheking in Abc
                if (check == false)
                {
                    myKey[0] = toupper((char) i);
                    myKey[1] = (char) j;
                    myKey[2] = (char) k;
                    check = checkNewHash(myKey);
                    if (check == true)
                    {
                        i = 123;
                        j = 123;
                        k = 123;
                    }
                }
//cheking in aBc
                if (check == false)
                {
                    myKey[0] = (char) i;
                    myKey[1] = toupper((char) j);
                    myKey[2] = (char) k;
                    check = checkNewHash(myKey);
                    if (check == true)
                    {
                        i = 123;
                        j = 123;
                        k = 123;
                    }
                }
//cheking in abC
                if (check == false)
                {
                    myKey[0] = (char) i;
                    myKey[1] = (char) j;
                    myKey[2] = toupper((char) k);
                    check = checkNewHash(myKey);
                    if (check == true)
                    {
                        i = 123;
                        j = 123;
                        k = 123;
                    }
                }
//cheking in ABC
                if (check == false)
                {
                    myKey[0] = toupper((char) i);
                    myKey[1] = toupper((char) j);
                    myKey[2] = toupper((char) k);
                    check = checkNewHash(myKey);
                    if (check == true)
                    {
                        i = 123;
                        j = 123;
                        k = 123;
                    }
                }
            }
        }
    }
    return check;
}

// ************************************************ check by 4 letter ********************************************************************
bool checkKey3(void)
{
    char myKey[5];
    memset(myKey, 0, sizeof(myKey));
    bool check = false;

    for (int i = 97; i <= 122; i++)
    {
        for (int j = 97; j <= 122; j++)
        {
            for (int k = 97; k <= 122; k++)
            {
                for (int t = 97; t <= 122; t++)
                {

//cheking abcd - all of them in lowercase
                    myKey[0] = (char) i;
                    myKey[1] = (char) j;
                    myKey[2] = (char) k;
                    myKey[3] = (char) t;
                    check = checkNewHash(myKey);
                    if (check == true)
                    {
                        i = 123;
                        j = 123;
                        k = 123;
                        t = 123;
                    }

//cheking in Abcd
                    if (check == false)
                    {
                        myKey[0] = toupper((char) i);
                        myKey[1] = (char) j;
                        myKey[2] = (char) k;
                        myKey[3] = (char) t;
                        check = checkNewHash(myKey);
                        if (check == true)
                        {
                            i = 123;
                            j = 123;
                            k = 123;
                            t = 123;
                        }
                    }
//cheking in aBcd
                    if (check == false)
                    {
                        myKey[0] = (char) i;
                        myKey[1] = toupper((char) j);
                        myKey[2] = (char) k;
                        myKey[3] = (char) t;
                        check = checkNewHash(myKey);
                        if (check == true)
                        {
                            i = 123;
                            j = 123;
                            k = 123;
                            t = 123;
                        }
                    }
//cheking in abCd
                    if (check == false)
                    {
                        myKey[0] = (char) i;
                        myKey[1] = (char) j;
                        myKey[2] = toupper((char) k);
                        myKey[3] = (char) t;
                        check = checkNewHash(myKey);
                        if (check == true)
                        {
                            i = 123;
                            j = 123;
                            k = 123;
                            t = 123;
                        }
                    }
//cheking in abcD
                    if (check == false)
                    {
                        myKey[0] = (char) i;
                        myKey[1] = (char) j;
                        myKey[2] = (char) k;
                        myKey[3] = toupper((char) t);
                        check = checkNewHash(myKey);
                        if (check == true)
                        {
                            i = 123;
                            j = 123;
                            k = 123;
                            t = 123;
                        }
                    }
//cheking in ABCD
                    if (check == false)
                    {
                        myKey[0] = toupper((char) i);
                        myKey[1] = toupper((char) j);
                        myKey[2] = toupper((char) k);
                        myKey[3] = toupper((char) t);
                        check = checkNewHash(myKey);
                        if (check == true)
                        {
                            i = 123;
                            j = 123;
                            k = 123;
                            t = 123;
                        }
                    }
                }
            }
        }
    }
    return check;
}
